### PR TITLE
irgen: Generate explicit instantiations

### PR DESCRIPTION
**Pull request by ChrisDodd**
- explicit instantiations for all Vector and IndexedVector used in any .def file

### CODE REVIEW

**Review by fruffy — COMMENTED** [review]
This seems to be a better approach than #4680 because it automatically applies to all IR nodes with these characteristics. 

I would add the comment/context from #4680 to the PR.

@vlstill can you confirm that this has a similar effect as your changes?

File: tools/ir-generator/irclass.cpp [hunk 1]
```diff
@@ -247,6 +259,8 @@ void IrDefinitions::generate(std::ostream &t, std::ostream &out, std::ostream &i
       << " inline bool operator!=(NodeDiscriminator lhs, RTTI::TypeId rhs) { return "
          "RTTI::TypeId(lhs) != rhs; }\n";
     t << "}  // namespace IR" << std::endl;
+
+    out << "#endif /* " << macroname << " */" << std::endl;
```
[the remark this answers is not in the document]
Comment (ChrisDodd) [reply tools/ir-generator/irclass.cpp:263]: No, I just moved it down from line 159, to keep it at the end of the generated file.

**Review by fruffy — APPROVED** [review]

**Comment by ChrisDodd** [comment]
> 
> Is it possible to move the declarations at the beginning, or is the ir-generator too much single-pass?

I'll try moving them earlier, but I think the instantiations need to be after the declarations of the classes that are used as parameters to the template?

**Comment by ChrisDodd** [comment]
> Is it possible to move the declarations at the beginning, or is the ir-generator too much single-pass?

Ok, I've moved the `extern template` declarations to the top of the ir-generated.h file -- this puts them before all the class definitions (and uses of the templates) -- see if that affects the build speed.

**Review by vlstill — APPROVED** [review]
Now the results are pretty much the same:

```
# this PR
  Time (mean ± σ):     566.911 s ±  3.272 s    [User: 11686.466 s, System: 588.923 s]
  Range (min … max):   562.361 s … 570.533 s    5 runs

# my approach
  Time (mean ± σ):     564.324 s ±  3.548 s    [User: 12161.493 s, System: 618.834 s]
  Range (min … max):   561.535 s … 570.435 s    5 runs
```
(I had to re-measure both as the PR base moved).

So the same. Thank you for this.

**Pull request merged**